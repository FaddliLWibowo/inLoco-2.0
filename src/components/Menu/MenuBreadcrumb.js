import React from 'react';

const MenuBreadcrumb = ({parentName}) => {
    return (
        <li className="menuBreadcrumb">{parentName ? parentName : 'Todas as camadas'}</li>
    );
}

export default MenuBreadcrumb;
