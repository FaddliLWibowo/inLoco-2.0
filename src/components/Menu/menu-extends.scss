%menu-item-icon {
    @extend %content-fa;
    font-size: 20px;
    margin-top: -5px;
    position: absolute;
}

%menu-item-arrow-right {
    @extend %menu-item-icon;
    content: $fa-var-angle-right; // >
    right: 21px;
}

%menu-item-arrow-left {
    @extend %menu-item-icon;
    content: $fa-var-angle-left; // <
    left: 21px;
}

%menu-item-checkbox-empty {
    @extend %menu-item-arrow-left;
    content: $fa-var-square-o;
    font-size: 14px;
    top: 50%;
    margin-top: -7px;
}

%menu-item-checkbox-checked {
    @extend %menu-item-checkbox-empty;
    content: $fa-var-check-square-o;
    color: $menu-selected-layer-color;
}

$menu-item-padding-vertical: 10px;
$menu-item-padding-horizontal: 17px;
$menu-item-padding-horizontal-with-arrow: 38px;

%padding-icon-right {
    padding: $menu-item-padding-vertical $menu-item-padding-horizontal-with-arrow $menu-item-padding-vertical $menu-item-padding-horizontal;
}
%padding-icon-left {
    padding: $menu-item-padding-vertical $menu-item-padding-horizontal $menu-item-padding-vertical $menu-item-padding-horizontal-with-arrow;
}
