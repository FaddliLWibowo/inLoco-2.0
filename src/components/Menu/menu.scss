%menu-item-icon {
    @extend %content-fa;
    font-size: 20px;
    margin-top: -5px;
    position: absolute;
}

%menu-item-arrow-right {
    @extend %menu-item-icon;
    content: $fa-var-angle-right; // >
    right: 21px;
}

%menu-item-arrow-left {
    @extend %menu-item-icon;
    content: $fa-var-angle-left; // <
    left: 21px;
}

%menu-item-checkbox-empty {
    @extend %menu-item-arrow-left;
    content: $fa-var-square-o;
    font-size: 14px;
    margin-top: -1px;
}

%menu-item-checkbox-checked {
    @extend %menu-item-checkbox-empty;
    content: $fa-var-check-square-o;
    color: $menu-selected-layer-color;
}

$menu-item-padding-vertical: 12px;
$menu-item-padding-horizontal: 17px;
$menu-item-padding-horizontal-with-arrow: 38px;

%padding-icon-right {
    padding: $menu-item-padding-vertical $menu-item-padding-horizontal-with-arrow $menu-item-padding-vertical $menu-item-padding-horizontal;
}
%padding-icon-left {
    padding: $menu-item-padding-vertical $menu-item-padding-horizontal $menu-item-padding-vertical $menu-item-padding-horizontal-with-arrow;
}

.sidebar-left > .menu {
    color: $menu-text-color;
    box-sizing: border-box;
    font-family: sans-serif;
    font-size: 11px;

    // fixes first element border
    > ul > li :first-child {
        border-top: 1px solid $menu-secondary-color;
    }

    // menu item
    li {

        @extend %padding-icon-right;

        list-style-type: none;
        cursor: pointer;
        border-bottom: 1px solid $menu-secondary-color;

        // menu items with children are uppercase
        &.menu-item-with-children, &.menu-item-all-layers {
            text-transform: uppercase;

            // arrow >
            &:after {
                @extend %menu-item-arrow-right;
            }
        }

        // the menu category that is opened, and the "all layers" menu item
        &.menu-item-all-layers, &.selected {
            background: $sidebar-left-highlight-color;
            color: $menu-breadcrumb-text-color;

            @extend %padding-icon-left;

            // arrow <
            &:after {
                @extend %menu-item-arrow-left;
            }
        }
    }

    // submenus
    ul {
        display: none;

        &.selected {
            display: block;

            li {
                @extend %padding-icon-left;

                &:after {
                    @extend %menu-item-checkbox-empty;
                }

                &.layer-selected {
                    color: $menu-selected-layer-color;
                    &:after {
                        @extend %menu-item-checkbox-checked;
                    }
                }
            }
        }
    }
}
