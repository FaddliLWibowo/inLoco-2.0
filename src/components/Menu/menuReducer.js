/**
* Creates menu array
* @param   {Object[]} layers Array of layers
* @returns {Object[]} Menu array with categories data, including layers IDs
*/
const menuReducer = (layers) => {
    let menu = []

    layers.forEach((layer) => {
        // creates top level menu item if it doesn't exists
        var menuTopLevelFound = false
        menu.forEach((menuItem) => {
            if (menuItem.id === layer.menu2[0]) {
                menuTopLevelFound = true
            }
        })
        if (!menuTopLevelFound) {
            if (layer.menu2 !== '' && layer.menu2[0].trim() !== '') {
                menu.push({
                    display: true,
                    id: layer.menu2[0],
                    title: layer.menu2[0],
                    layers: [],
                    idMenu: menu.length,
                    subMenu: []
                })
            }
        }

        // check if layer has submenu
        if (layer.menu2 !== '' && layer.menu2[1] && layer.menu2[1].trim() !== '') {

            // find menu element and add submenu to it
            menu.forEach((menuItem) => {
                if (menuItem.id === layer.menu2[0] ){

                    // found menu element, checking for submenu
                    let subMenuFound = false
                    menuItem.subMenu.forEach((subMenuItem) => {
                        if (subMenuItem.id === layer.menu2[1]) {
                            subMenuFound = true
                        }
                    })
                    if (!subMenuFound) {
                        menuItem.subMenu.push({
                            display: true,
                            id: layer.menu2[1],
                            title: layer.menu2[1],
                            layers: [],
                            idMenu: menuItem.subMenu.length
                        })
                    }
                }
            })
        }

        // then add the layer ID to an array of its menu item
        menu.forEach((menuItem) => {
            if (menuItem.id === layer.menu2[0]) {
                // first check if we should add to submenu
                if (layer.menu2[1] && menuItem.subMenu.length > 0) {
                    menuItem.subMenu.forEach((subMenuItem) => {
                        if (subMenuItem.id === layer.menu2[1]) {
                            subMenuItem.layers.push(layer.key)
                        }
                    })
                }
                // if not, add to menu
                else {
                    menuItem.layers.push(layer.key)
                }
            }
        })
    })

    // finally, sort menu categories in A-Z
    menu.sort((a, b)=>{
        if (a.title < b.title) {
            return -1
        }

        if (a.title > b.title) {
            return 1
        }

        return 0
    })

    return menu
}

export default menuReducer
