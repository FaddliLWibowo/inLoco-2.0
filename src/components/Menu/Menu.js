import React from 'react';
import MenuItem from '../MenuItem/MenuItem';

const Menu = ({menuItems, layers, onLayerClick, onMenuClick}) => {
    return (
        <ul className="menu">
            {
                menuItems.map((item) => <MenuItem item={item} layers={layers} onLayerClick={onLayerClick} onMenuClick={onMenuClick} onItemClick={Number.isInteger(item) ? onLayerClick : onMenuClick} key={Number.isInteger(item) ? item : item.idMenu} />)
            }
        </ul>
    );
}

export default Menu;
