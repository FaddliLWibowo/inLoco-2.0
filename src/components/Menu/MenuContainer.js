import React from 'react'
import Menu from './Menu'
import { connect } from 'react-redux'
import { toggleLayer, toggleMenu, untoggleAll, showDescription, hideDescription } from '../../actions/actions.js'

const getVisibleMenuElements = (menuItems) => {
    let hasOneVisibleMenuItem = false
    for (let menuItem of menuItems) {
        if (menuItem.selected) {
            hasOneVisibleMenuItem = true
        }
    }

    if (!hasOneVisibleMenuItem) {
        return menuItems
    }

    return menuItems.filter(menuItem => menuItem.selected)
}

const getVisibleLayers = (layers) => {
    let hasOneVisibleLayer = false
    for (let layer of layers) {
        if (layer.match) {
            hasOneVisibleLayer = true
        }
    }

    if (!hasOneVisibleLayer) {
        return layers
    }

    return layers.filter(layer => layer.match)
}

const mapStateToProps = (state, ownProps) => {
    return {
        menuItems: Array.isArray(state.menuItems) ? state.menuItems : [],
        layers: Array.isArray(state.layers) ? state.layers : [],
        currentLevel: state.currentLevel,
        sidebarLeftWidth: ownProps.sidebarLeftWidth,
        sidebarLeftHeight: ownProps.sidebarLeftHeight,
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        onLayerClick: (item) => {
            dispatch(toggleLayer(item))
        },
        onMenuItemClick: (item) => {
            // scroll to top
            document.getElementsByClassName('sidebar-left')[0].childNodes[1].scrollTop = 0

            dispatch(toggleMenu(item))
        },
        onUntoggleAllClick: () => {
            dispatch(untoggleAll())
        },
        onMouseOver: (layer,
            sidebarLeftWidth,
            parentHeight,
            top
        ) => {
            if(layer){
                dispatch(showDescription(
                    layer,
                    sidebarLeftWidth,
                    parentHeight,
                    top
                ))
            }
        },
        onMouseOut: (layer) => {
            if(layer){
                dispatch(hideDescription(layer))
            }
        }
    };
};

const MenuContainer = connect(
    mapStateToProps,
    mapDispatchToProps
)(Menu)

export default MenuContainer
