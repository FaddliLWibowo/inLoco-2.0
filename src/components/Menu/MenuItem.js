import React from 'react'
import Menu from './Menu'

const MenuItem = ({item, layers, onItemClick, onMouseOver, sidebarLeftWidth, sidebarLeftHeight, onMouseOut, onMenuItemClick, onLayerClick, parentMenuTitle, currentLevel, allMenuItems}) => {
    // class name if menu item with children or single layer, with no children
    let menuItemClassName = item.title ? 'menu-item-with-children' : 'menu-layer'
    menuItemClassName += item.selected ? ' selected' : ''

    let visibleClass = ''
    let itemTitle = ''

    // check if one menu is selected, and no other items are matched, hide others
    let otherIsSelected = false
    let otherIsNotMatched = false
    if (allMenuItems) {
        for (let otherMenuItem of allMenuItems) {
            if (otherMenuItem.title && otherMenuItem.title !== item.title) {
                if (otherMenuItem.selected) {
                    otherIsSelected = true
                }
                if (!otherMenuItem.match) {
                    otherIsNotMatched = true
                }
            }
        }
        if (otherIsSelected && !otherIsNotMatched) {
            // we can only hide other items if not on search
            visibleClass = 'hidden'
        }
    }

    // if menu with children, and it doesn't match search, hide it
    if (item.title) {
        itemTitle = item.title
        if (!item.match) {
            visibleClass = 'hidden'
        }
    } else {
        // if it's a layer, check if it's match'ed.
        for (var i = 0 ; i < layers.length ; i++) {
            var layer = layers[i]
            if (layer.key === item) {
                itemTitle = layer.title
                if (!layer.match) {
                    visibleClass = 'hidden'
                }
            }
        }
    }
    visibleClass += " menu-item-container"

    // check if layer is selected
    if (!item.title){
        for (var i = 0 ; i < layers.length ; i++) {
            var layer = layers[i]
            if (layer.key === item && layer.selected) {
                menuItemClassName += ' layer-selected';
            }
        }
    }

    return (
        <div className={visibleClass}>
            <li
                onMouseOut={() => onMouseOut(item.id ? undefined : layers[item])}
                onMouseOver={(event) => onMouseOver(event, item.id ? undefined : layers[item], sidebarLeftWidth, sidebarLeftHeight)}
                onClick={() => onItemClick(item.id ? item : layers[item])}
                className={menuItemClassName}
            >
                { itemTitle }
            </li>
            {
                item.layers ?
                <Menu
                    menuItems={item.layers}
                    menuTitle={item.title}
                    parentMenuTitle={parentMenuTitle}
                    key={item.idMenu}
                    selected={item.selected}
                    layers={layers}
                    onMenuItemClick={onMenuItemClick}
                    onLayerClick={onLayerClick}
                    onMouseOver={onMouseOver}
                    sidebarLeftWidth={sidebarLeftWidth}
                    sidebarLeftHeight={sidebarLeftHeight}
                    onMouseOut={onMouseOut}
                    currentLevel={currentLevel}
                />
                : ''
            }
        </div>
    );
}

export default MenuItem
