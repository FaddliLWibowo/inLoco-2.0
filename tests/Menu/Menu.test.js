import React from 'react';
import Menu from '../../src/components/Menu/Menu.js';

it('Menu component renders correctly with no data', () => {
    const component = shallow(<Menu/>);
    expect(component).toMatchSnapshot();
});

it('Menu component renders correctly with real data', () => {
    let mock = {
        "sidebarLeftWidth":37.859375,
        "sidebarLeftHeight":85.046875,
        "menuItems":[
            {
                "display":true,
                "id":"Assistência Social",
                "title":"Assistência Social",
                "layers":[1],
                "idMenu":1,
                "isSubMenu":false,
                "submenus":[],
                "selected":false,
                "match":true
            }
        ],
        "layers":[
            {
                "id":"plataforma_acterj_conselho_tutelar",
                "name":"acterj_conselho_tutelar",
                "title":"Conselho Tutelar",
                "workspace":"plataforma",
                "display":true,
                "restricted":false,
                "layerName":"plataforma:acterj_conselho_tutelar",
                "description":"Fontes: Associação dos Conselheiros Tutelares (ACTERJ). Contato: acterj@gmail.com/ http://www.acterj.org.br/ \r\nAno:2016.",
                "bbox":"-44.71388499999997,-23.223640999999986,-41.045743999999964,-20.96493900000002",
                "caops":["infancia","cidadania","educacao"],
                "menu":"assistencia",
                "menu2":["Assistência Social"],
                "key":0,
                "styles":[
                    {
                        "id":0,
                        "title":"Conselho Tutelar",
                        "name":"plataforma:conselho_tutelar",
                        "description":"Conselhos Tutelares registrados na ACTERJ e CEDCA",
                        "icon":"/geoserver/plataforma/wms?service=WMS&version=1.1.0&request=GetMap&bbox=-44.71388499999997,-23.223640999999986,-41.045743999999964,-20.96493900000002&width=27&height=20&ssrs=EPSG:3857&format=image%2Fpng&layers=plataforma:acterj_conselho_tutelar&singleTile=true&styles=plataforma:conselho_tutelar",
                        "icon_2":"./resources/img/plataforma/icones/plataforma_acterj_conselho_tutelar-plataforma_conselho_tutelar_27_20.png","thumb":"/geoserver/plataforma/wms?service=WMS&version=1.1.0&request=GetMap&bbox=-44.71388499999997,-23.223640999999986,-41.045743999999964,-20.96493900000002&width=225&height=150&ssrs=EPSG:3857&format=image%2Fpng&layers=plataforma:acterj_conselho_tutelar&singleTile=true&styles=plataforma:conselho_tutelar",
                        "thumb_2":"./resources/img/plataforma/icones/plataforma_acterj_conselho_tutelar-plataforma_conselho_tutelar_225_150.png","tooltip":"<span style=\"width:750px;\"><img src=\"./resources/img/plataforma/icones/plataforma_acterj_conselho_tutelar-plataforma_conselho_tutelar_225_150.png\" onError=\"this.onerror=null;this.src='/geoserver/plataforma/wms?service=WMS&version=1.1.0&request=GetMap&bbox=-44.71388499999997,-23.223640999999986,-41.045743999999964,-20.96493900000002&width=225&height=150&ssrs=EPSG:3857&format=image%2Fpng&layers=plataforma:acterj_conselho_tutelar&singleTile=true&styles=plataforma:conselho_tutelar'\" width=\"150\" class=\"estilo-tooltip\"/><strong>Conselho Tutelar</strong><br>Conselho Tutelar</span>"
                    }
                ],
                "selected":false,
                "match":true,
                "showDescription":false,
                "selectedLayerStyleId":0
            }
        ],
        "currentLevel":0
    }

    const component = shallow(<Menu/>);
    expect(component).toMatchSnapshot();
});
