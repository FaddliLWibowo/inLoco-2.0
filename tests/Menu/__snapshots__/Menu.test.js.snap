// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu component renders correctly 1`] = `
<ul
  className="menu"
/>
`;
